### PR TITLE
OG-427 Make Relay Server account for 63/64 rule estimating 'relayCall' gas limit

### PR DESCRIPTION
**Pull request by forshtat**
1. Emit 'gasleft' in the TestRecipient to check it is close to requested
2. Set high target worker balance for large gas limit test
3. Take into account calldata cost when calculating external gas limit
(defaultEnvironment is hard-coded as of now)
4. Multiply gas limit by a factor to accomodate EIP-150 63/64th rule
(factor hardcoded at 1.1 as of now)
5. Increase 'run-with-testrpc' gas limit for large gaslimit test

### CODE REVIEW

File: src/relayserver/RelayServer.ts [hunk 1]
```diff
@@ -40,6 +40,18 @@ import { toBuffer } from 'ethereumjs-util'
 
 import Timeout = NodeJS.Timeout
 
+/**
+ * After EIP-150, every time the call stack depth is increased without explicit call gas limit set,
+ * the 63/64th rule is applied to gas limit.
+ * As we have to pass enough gas to a transaction to pass 'relayRequest.request.gas' to the recipient,
+ * and this check is at stack depth of 3, we have to oversupply gas to an outermost ('relayCall') transaction
+ * by approximately 1/(63/64)^3 times.
```
Comment (drortirosh) [inline src/relayserver/RelayServer.ts:48]: yes, this the approximation, but its 1.05, not 1.1

File: contracts/test/TestRecipient.sol [hunk 2]
```diff
@@ -57,7 +58,7 @@ contract TestRecipient is BaseRelayRecipient {
     function dontEmitMessage(string calldata message) public {}
 
     function emitMessageNoParams() public {
-        emit SampleRecipientEmitted("Method with no parameters", _msgSender(), msg.sender, tx.origin, 0, address(this).balance);
+        emit SampleRecipientEmitted("Method with no parameters", _msgSender(), msg.sender, tx.origin, 0, gasleft(), address(this).balance);
```
[the remark this answers is not in the document]
Comment (forshtat) [reply contracts/test/TestRecipient.sol:61]: Do you mean rename it or do you mean get `gasleft` before emit? because here it does not make much sense, this method doesn't do anything else, just the emit.

File: src/relayserver/RelayServer.ts [hunk 3]
```diff
@@ -40,6 +40,18 @@ import { toBuffer } from 'ethereumjs-util'
 
 import Timeout = NodeJS.Timeout
 
+/**
+ * After EIP-150, every time the call stack depth is increased without explicit call gas limit set,
+ * the 63/64th rule is applied to gas limit.
```
[the remark this answers is not in the document]
Comment (forshtat) [reply src/relayserver/RelayServer.ts:45]: Is there something I should do about it? The comment says "without explicit call gas limit".

File: test/Flows.test.ts [hunk 4]
```diff
@@ -157,6 +163,29 @@ options.forEach(params => {
...
+            it(`should calculate valid external tx gas limit for a transaction with inner call gas limit of  ${innerGasLimit.toString()}`, async function () {
+              const gas = innerGasLimit
+              let res: any
+              try {
+                res = await sr.emitMessageNoParams({ from, gas })
+              } catch (e) {
+                console.log('error is ', e.message)
+                throw e
+              }
+              const actual: BN = res.logs[0].args.gasLeft
```
[the remark this answers is not in the document]
Comment (forshtat) [reply test/Flows.test.ts]: Fixed

File: package.json [hunk 5]
```diff
@@ -35,8 +35,8 @@
   "scripts": {
     "test": "yarn run lint && yarn generate && yarn tsc && npm run test-js",
     "ganache": "ganache-cli --chainId 1337 --hardfork 'istanbul' --gasLimit 100000000 --defaultBalanceEther 1000 --deterministic --keepAliveTimeout 2147483647",
-    "test-js": "run-with-testrpc --chainId 1337 --hardfork 'istanbul' --port 8544 --gasLimit 10000000 --defaultBalanceEther 1000 --deterministic 'npx truffle --network npmtest test'",
-    "test-bail": "run-with-testrpc --chainId 1337 --hardfork 'istanbul' --port 8544 --gasLimit 10000000 --defaultBalanceEther 1000 --deterministic 'npx truffle --network npmtest test --bail'",
+    "test-js": "run-with-testrpc --chainId 1337 --hardfork 'istanbul' --port 8544 --gasLimit 100000000 --defaultBalanceEther 1000 --deterministic 'npx truffle --network npmtest test'",
```
[the remark this answers is not in the document]
Comment (forshtat) [reply package.json]: you are actually right, removed the 90M gas limit test and limited ganache gas to 12M

File: src/relayserver/RelayServer.ts (hunk 1 above)
Comment (forshtat) [reply src/relayserver/RelayServer.ts:48]: I can live with it. It is better to oversupply then to catch a forwarder revert.

File: contracts/test/TestRecipient.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (drortirosh) [reply contracts/test/TestRecipient.sol:61]: not really critical.

File: test/Flows.test.ts [hunk 6]
```diff
@@ -157,6 +163,29 @@ options.forEach(params => {
       let approvalPaymaster: TestPaymasterPreconfiguredApprovalInstance
       let relayProvider: RelayProvider
 
+      /**
+       * RelayServer has to oversupply gas in order to pass over all the 'remaining gas' checks and give enough gas
+       * to the recipient. We are setting 10M as a block gas limit, so cannot go much higher than 9M gas here.
+       */
+      describe('with different gas limits', function () {
```
Comment (drortirosh) [inline test/Flows.test.ts:170]: I remember I wrote a comment, but I don't see it here:

this test doesn't belong here: Flows are for normal sanity flows, not for edge-cases.
this test belongs to the gas calculations suite.

Comment (forshtat) [reply test/Flows.test.ts:170]: this only really fails on an integration test, and this is the file with integration tests right now

**Review by drortirosh — APPROVED** [review]
